File: 2-AddTwoNumbers/2-AddTwoNumbers.cpp
// Last updated: 6/11/2026, 11:38:51 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* node){
        ListNode* cur = node;
        ListNode* next;
        ListNode* prev = nullptr;

        while(cur){
            next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }

        return prev;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        l1 = l1;
        l2 = l2;

        ListNode* t1 = l1;
        ListNode* t2 = l2;

        ListNode* newh = new ListNode(0);
        ListNode* t= newh;

        int carry = 0;
        while(t1 && t2){
            int actsum = t1->val + t2->val + carry;
            t->next = new ListNode(actsum%10);
            carry = actsum/10;
            t1 = t1->next;
            t2 = t2->next;
            t = t->next;
        }

        while(t1){
            int actsum = t1->val+carry;
            t->next = new ListNode(actsum%10);
            t1 = t1->next;
            t = t->next;
            carry = actsum/10;
        }

        while(t2){
            int actsum = t2->val + carry;
            t->next = new ListNode(actsum%10);
            t2 = t2->next;
            t = t->next;
            carry = actsum/10;
        }

        if(carry){
            t->next = new ListNode(carry);
            t = t->next;
        }

        return newh->next;


    }
};

File: 2-AddTwoNumbers/2-AddTwoNumbers.py
# Last updated: 6/11/2026, 11:38:56 AM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        sum1 = 0
        carry = 0
        l3 = ListNode()
        head = l3
        while(l1!=None or l2!=None or carry!=0):
            if(l1 != None):
                sum1+= l1.val
                l1 = l1.next
            if(l2 != None):
                sum1+= l2.val
                l2 = l2.next
            sum1 += carry
            carry = sum1//10
            n = sum1%10

            n1 = ListNode(n)
            head.next = n1
            head = n1  

            sum1 = 0       
            
        return l3.next
